Feature/keystone auth

chianingwang opened the pull request:
Add support keystone auth v2 and v3 for cloud sync.
- s3_sync/sync_swift.py (m)
- test/unit/test_sync_swift.py (m)

timuralp commented:
@chianingwang could you rebase on master, now that we have travis hooked up, so we can run the unit/integration/flake8 tests?

chianingwang commented:
@timuralp I rebase and it looks like auto trigger the Travis test job but it looks like fail which can't get the container ID. I might miss something and might consult you tomorrow when I am in SF office.
```
$ container_id=$(mktemp)
before_install.3
0.27s
$ docker run -d -v `pwd`:/swift-s3-sync cloud-sync/test > "${container_id}"
6.64s
$ docker exec -it `cat ${container_id}` timeout 30 bash -c 'until echo > /dev/tcp/localhost/8080; do sleep 0.5; done'
bash: connect: Cannot assign requested address
bash: /dev/tcp/localhost/8080: Cannot assign requested address```

chianingwang commented:
@timuralp I found the error and fix it. They are a most likely pem8 format issues.

One thing I don't know how to deal with which is as below.
Here is original code but it complains `'project_domain_name': self.settings['project_domain_name'],` line length 80 is longer than limitation 79.
```
            elif authtype == 'keystone_v3':
                _os_options = {
                  'user_domain_name': self.settings['user_domain_name'],
                  'project_domain_name': self.settings['project_domain_name'],
                  'project_name': self.settings['project_name']}
```

Thus I have to update the code like as below to deal with this issue.
```
            elif authtype == 'keystone_v3':
                userdomainname = self.settings['user_domain_name']
                projectdomainname = self.settings['project_domain_name']
                projectname = self.settings['project_name']
                _os_options = {
                    'user_domain_name': userdomainname,
                    'project_domain_name': projectdomainname,
                    'project_name': projectname}
```

I'm not sure whether there has a better way to handle this, if yes, pleases advice.

timuralp commented:
@chianingwang I think that's fine. Alternatively, if you rewrite using the `dict` syntax, you could save some characters, as well. For example:
```
_os_options = dict(
    user_domain_name=self.settings['user_domain_name'],
    project_domain_name=self.settings['project_domain_name'],
    project_name=self.settings['project_name']
```

Of course, since the fields map in the config file map exactly to the code, you could even do this:
```
os_auth_fields = ['project_name', 'project_domain_name', 'user_domain_name']
_os_options = dict([(field, self.settings[field]) for field in os_auth_fields])
```

I don't have a strong preference around this, so you can pick whichever you prefer.

At the high level, can you please squash your commits and remove any commented out code? I hesitate to review this PR in its current state. Once that's done, I can start looking at the code.

chianingwang commented:
@timuralp Thanks for your advice, I took the 2nd advice from you which I'm doing as below and looks good and pass the `Travis CI`
```
            elif authtype == 'keystone_v3':
                _os_auth_fields = ['project_name',
                                              'project_domain_name',
                                              'user_domain_name']
                _os_options = dict([(field, self.settings[field])
                                               for field in _os_auth_fields])
```
Moreover, I squash your commits and remove any commented out code.
Now the commit is only as below. I can squeeze them more or remove those comment totally. 
```
        # **connection_args is common arguments; authurl, user, and key
        common_connection_kwargs = {'authurl': endpoint,
                                    'user': username,
                                    'key': key,
                                    'retries': 3}

        # default: swift, options keystone_v2 and keystone_v3
        authtype = self.settings.get('auth_type', None)
        if authtype is None:
            authtype = 'swift'
```

timuralp commented:
@chianingwang thanks for cleaning up some of the comment stuff! Can you squash the actual commits, please? There are currently 19 commits in this PR. Ideally, we would have one commit that implements the change required to support the keystone authentication option. Let me know if squashing commits with git is confusing to you -- happy to point you in the right direction!

timuralp commented on s3_sync/sync_swift.py:
```diff
@@ -48,9 +48,40 @@ def _get_client_factory(self):
-            return swiftclient.client.Connection(
-                authurl=endpoint, user=username, key=key, retries=3)
+            if _authtype == 'keystone_v2':
+                _tenant_name = self.settings['tenant_name']
```
Consider changing this to a `.get()`. This way we won't hit a `KeyError`

timuralp commented on s3_sync/sync_swift.py:
```diff
@@ -48,9 +48,40 @@ def _get_client_factory(self):
+                        tenant_name=_tenant_name,
+                        **common_connection_kwargs)
+                else:
+                    raise NotImplementedError()
```
Is this the right exception to raise? We do implement support for this auth type, but are missing arguments.

timuralp commented on s3_sync/sync_swift.py:
```diff
@@ -48,9 +48,40 @@ def _get_client_factory(self):
+        # default: swift, options keystone_v2 and keystone_v3
+        _authtype = self.settings.get('auth_type', None)
+        if _authtype is None:
+            _authtype = 'swift'
```
Why not put `'swift'` as the argument to `.get()` on line 58?

timuralp commented on test/unit/test_sync_swift.py:
```diff
@@ -36,7 +36,8 @@ def setUp(self):
              'container': 'container',
-             'aws_endpoint': 'http://swift.url/auth/v1.0'})
+             'aws_endpoint': 'http://swift.url/auth/v1.0',
+             'auth_type': 'swift'})
```
Is this necessary? A missing argument should default to `swift` anyway?

timuralp commented on test/unit/test_sync_swift.py:
```diff
@@ -556,6 +557,7 @@ def test_per_account_bucket(self, mock_swift):
              'aws_secret': 'credential',
              'account': 'account',
              'container': 'container',
+             'auth_type': 'swift',
```
Same question as above.

timuralp commented on test/unit/test_sync_swift.py:
```diff
@@ -581,3 +583,43 @@ def test_per_account_container_create(self, mock_swift):
+    def test_keystone_v2_auth(self, mock_swift):
+        mock_swift.return_value = mock.Mock()
+
+        # in this case, the "bucket" is actually the prefix
```
I think this is an out of place artifact from a copy/paste.

timuralp commented on test/unit/test_sync_swift.py:
```diff
@@ -581,3 +583,43 @@ def test_per_account_container_create(self, mock_swift):
+             'aws_endpoint': 'http://swift.url/auth/v1.0'},
+            per_account=True)
+
+        self.assertEqual('sync_container', sync_swift.remote_container)
```
You should assert that the `swiftclient.client.Connection` constructor was called with the expected arguments. Also, why do we want to assert the container name?

timuralp commented on test/unit/test_sync_swift.py:
```diff
@@ -581,3 +583,43 @@ def test_per_account_container_create(self, mock_swift):
+    def test_keystone_v3_auth(self, mock_swift):
+        mock_swift.return_value = mock.Mock()
+
+        # in this case, the "bucket" is actually the prefix
```
Same comments as on the above test.

timuralp commented on test/unit/test_sync_swift.py:
```diff
@@ -581,3 +583,43 @@ def test_per_account_container_create(self, mock_swift):
+             'aws_endpoint': 'http://swift.url/auth/v1.0'},
+            per_account=True)
+
+        self.assertEqual('sync_container', sync_swift.remote_container)
```
Same comments about the assertion.

chianingwang commented:
@timuralp, Thanks for helping me the code review and those are all great advice! I modified the code and commit it. Please help to review it and any advice will be welcome.

timuralp commented on s3_sync/sync_swift.py:
```diff
@@ -48,9 +48,41 @@ def _get_client_factory(self):
+                        auth_version='3',
+                        os_options=_os_options,
+                        **common_connection_kwargs)
+                except Exception as e:
```
This is confusing to me. Why do we rescue the exception and then re-raise it? It seems that this could be removed entirely?

timuralp commented on s3_sync/sync_swift.py:
```diff
@@ -48,9 +48,41 @@ def _get_client_factory(self):
+                        auth_version='2',
+                        tenant_name=_tenant_name,
+                        **common_connection_kwargs)
+                except Exception as e:
```
This is confusing to me. Why do we rescue the exception and then re-raise it? It seems that this could be removed entirely?

timuralp commented on s3_sync/sync_swift.py:
```diff
@@ -48,9 +48,41 @@ def _get_client_factory(self):
+            if _authtype == 'keystone_v2':
+                try:
+                    _tenant_name = self.settings['tenant_name']
+                    return swiftclient.client.Connection(
```
Ok, re-reading this, I think we can clean up the code a little more. In the `if` statements, we should only mutate the `common_connection_kwargs` dictionary (not actually calling `swiftclient.client.Connection`). And then have exactly one call to the method on line 85.

timuralp commented on s3_sync/sync_swift.py:
```diff
@@ -48,9 +48,41 @@ def _get_client_factory(self):
+                    raise e
+            else:
+                return swiftclient.client.Connection(
+                    authurl=endpoint, user=username, key=key, retries=3)
```
This is exactly `common_connection_kwargs`. In fact, maybe naming it `connection_kwargs` (or `connection_args`) would be better.

timuralp commented on test/unit/test_sync_swift.py:
```diff
@@ -581,3 +581,75 @@ def test_per_account_container_create(self, mock_swift):
+        missing_tenant_name = self.sync_swift._get_client_factory()
+        with self.assertRaises(Exception) as context:
+            missing_tenant_name()
+        self.assertTrue('tenant_name' in str(context.exception))
```
Could we have a somewhat more specific message to check in the exception?

timuralp commented on test/unit/test_sync_swift.py:
```diff
@@ -581,3 +581,75 @@ def test_per_account_container_create(self, mock_swift):
+             'aws_endpoint': 'http://swift.url/auth/v1.0'},
+            per_account=True)
+        missing_tenant_name = self.sync_swift._get_client_factory()
+        with self.assertRaises(Exception) as context:
```
Is there a more specific exception than `Exception`?

timuralp commented on test/unit/test_sync_swift.py:
```diff
@@ -581,3 +581,75 @@ def test_per_account_container_create(self, mock_swift):
+             'aws_endpoint': 'http://swift.url/auth/v1.0'},
+            per_account=True)
+        missing_project_domain_name = self.sync_swift._get_client_factory()
+        with self.assertRaises(Exception) as context:
```
We should have a more specific exception to check here (than `Exception`).

timuralp commented on test/unit/test_sync_swift.py:
```diff
@@ -581,3 +581,75 @@ def test_per_account_container_create(self, mock_swift):
+             'aws_endpoint': 'http://swift.url/auth/v1.0'},
+            per_account=True)
+        missing_project_name = self.sync_swift._get_client_factory()
+        with self.assertRaises(Exception) as context:
```
We should have a more specific exception to check here (than `Exception`).

timuralp commented on test/unit/test_sync_swift.py:
```diff
@@ -581,3 +581,75 @@ def test_per_account_container_create(self, mock_swift):
                          swift_client.mock_calls)
+
+    @mock.patch('s3_sync.sync_swift.swiftclient.client.Connection')
+    def test_keystone_v2_auth_missing_argument(self, mock_swift):
```
At the high level, can you write a test that asserts that the correct parameters are passed to the Connection() constructor? As is, we only have negative tests (tests that test for failure).

timuralp commented on test/unit/test_sync_swift.py:
```diff
@@ -581,3 +581,75 @@ def test_per_account_container_create(self, mock_swift):
+        self.assertTrue('tenant_name' in str(context.exception))
+
+    @mock.patch('s3_sync.sync_swift.swiftclient.client.Connection')
+    def test_keystone_v3_auth_missing_arguments(self, mock_swift):
```
This test has a lot of copy/pasted code. Can you improve on that? For example, the only thing that changes between each invocation is the missing parameter, right? We also need a test that verifies that when all parameters are present, we call the `Connection` class with the right arguments. Can you please write such a test?

chianingwang commented:
@timuralp  Great input and really appreciated the code review.
I modified the codes follow your suggestions but not sure I cover all of them.
Here is the summary what I changed. Please share your thought and any comment will be welcome.

##### s3_sync/sync_swift.py
 * rename `common_connection_kwargs` to `connection_kwargs`
 * handle `connection_kwargs` dict first and add required arguments base on `auth_type`
 * return connection at the end against `connection_kwargs`
 * more speficifc exception message via `raise Excpetion` 

##### test/unit/test_sync_swift.py
 * test_keystone_v2_auth_missing_tenant_name
 * test_keystone_v3_auth_missing_project_name
 * test_keystone_v3_auth_missing_project_domain_name
 * test_keystone_v3_auth_missing_user_domain_name
 * test_keystone_v2_auth_current_withouterror
 * test_keystone_v3_auth_current_withouterror
 
ps: I squash all the commits into 1 ! :+1:

chianingwang commented:
@timuralp I followed your advice and practice it, and run the unit test to test it. It's all good for me.

timuralp commented:
Closing this PR as we merged code based on it already and have keystone support. Thanks for spearheading this, @chianingwang!
